### PR TITLE
async-coap-uri: Fix macro hygiene

### PR DESCRIPTION
**Pull request by Luro02**
There were some places that expected the type/trait to be imported and other parts did not.

For example:

```rust
macro_rules! _impl_uri_buf_traits_base {
    ( $C:ty , $B:ty ) => {
        _impl_uri_traits!($C);

        impl core::convert::From<$C> for std::string::String {
            fn from(x: $C) -> Self {
                String::from(x.0)
            }
        }
        // ...
    };
    // ...
}
```
where the `String` has an absolute path, but the `String` in the function body does not have an absolute path.

I also changed relative paths (`core::convert::From`) to absolute paths (`::core::convert::From`) to prevent naming conflicts (e.g. declaring a `core` module).

---

The second part of this PR fixes an issue, when importing this crate's macros:

```rust
use async_coap_uri::uri;

let uri = uri!("https://www.example.com/");
```

would error with

```
error: cannot find macro `assert_uri_literal` in this scope
 --> src/main.rs:4:15
  |
4 |     let uri = uri!("https://www.example.com/");
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

error: could not compile `uri`.

To learn more, run the command again with --verbose
```

I actually had no clue how to solve this, because prepending `$crate::` to the `assert_uri_literal` macro call caused another error that could only be solved by reverting the change.

So I asked in the forum https://users.rust-lang.org/t/macro-rules-resolution-hell/40236 and this is the result :)

TL;DR: `async-coap-uri` should not use its own macros, otherwise all downstream crates have to import internal macros. This can be solved by creating 2 macros (one is identical to the old one and the other has the above mentioned fix applied).

### CODE REVIEW

**Comment by Luro02** [comment]
@darconeous would you mind reviewing this PR? 🤗

**Comment by darconeous** [comment]
I wonder if there is a way to add a `cfg` that we could use to differentiate between the case where we are compiling the crate vs just using the crate?
